crates/core/src/table/file_format_options.rs

```diff
@@ -0,0 +1,308 @@
+    fn compression(&self, column_path: &ColumnPath) -> Compression;
+    async fn create_writer_properties(
+        &self,
+        file_path: &Path,
+        file_schema: &Arc<ArrowSchema>,
+    ) -> DeltaResult<WriterProperties>;
+}
+
+#[derive(Clone, Debug)]
+pub struct SimpleWriterPropertiesFactory {
+    writer_properties: WriterProperties,
+}
+
+impl SimpleWriterPropertiesFactory {
+    pub fn new(writer_properties: WriterProperties) -> Self {
+        Self { writer_properties }
+    }
+}
+
+impl Default for SimpleWriterPropertiesFactory {
+    fn default() -> Self {
+        let writer_properties = WriterProperties::builder()
+            .set_compression(Compression::SNAPPY) // Code assumes Snappy by default
+            .set_created_by(format!("delta-rs version {}", crate_version()))
+            .build();
+        Self { writer_properties }
+    }
+}
+
+#[async_trait]
+impl WriterPropertiesFactory for SimpleWriterPropertiesFactory {
+    fn compression(&self, column_path: &ColumnPath) -> Compression {
+        self.writer_properties.compression(column_path)
+    }
+
+    async fn create_writer_properties(
+        &self,
+        file_path: &Path,
+        _file_schema: &Arc<ArrowSchema>,
+    ) -> DeltaResult<WriterProperties> {
+        info!("Called create_writer_properties for file: {file_path}");
+        Ok(self.writer_properties.clone())
+    }
+}
+
+// More advanced factory with KMS support
+#[cfg(feature = "datafusion")]
+#[derive(Clone, Debug)]
+pub struct KMSWriterPropertiesFactory {
+    writer_properties: WriterProperties,
+    encryption: Option<crate::operations::encryption::TableEncryption>,
+}
+
+#[cfg(feature = "datafusion")]
+impl KMSWriterPropertiesFactory {
+    pub fn with_encryption(table_encryption: TableEncryption) -> Self {
+        let writer_properties = WriterProperties::builder()
+            .set_compression(Compression::SNAPPY) // Code assumes Snappy by default
+            .set_created_by(format!("delta-rs version {}", crate_version()))
+            .build();
+        Self {
+            writer_properties,
+            encryption: Some(table_encryption),
+        }
+    }
+}
+
+#[cfg(feature = "datafusion")]
+#[async_trait]
+impl WriterPropertiesFactory for KMSWriterPropertiesFactory {
+    fn compression(&self, column_path: &ColumnPath) -> Compression {
+        self.writer_properties.compression(column_path)
+    }
+
+    async fn create_writer_properties(
+        &self,
+        file_path: &Path,
+        file_schema: &Arc<ArrowSchema>,
+    ) -> DeltaResult<WriterProperties> {
+        let mut builder = self.writer_properties.to_builder();
+        if let Some(encryption) = self.encryption.as_ref() {
+            builder = encryption
+                .update_writer_properties(builder, file_path, file_schema)
+                .await?;
+        }
+        Ok(builder.build())
+    }
+}
+
+/// AI generated code to get builder from existing WriterProperties
```

Comment:
We should put a link to the arrow-rs PR that will allow removing this (https://github.com/apache/arrow-rs/pull/8272)